packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1922 @@
+    origin: OriginString,
+    targetName: ExtractPermission<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >['parentCapability'],
+  ): void {
+    const { allowedCaveats, validator } = specification;
+    if (permission.caveats !== null) {
+      permission.caveats.forEach(({ type }) => {
+        if (!allowedCaveats?.includes(type)) {
+          throw new ForbiddenCaveatError(type, origin, targetName);
+        }
+      });
+    }
+
+    if (validator) {
+      validator(permission, origin, targetName);
+    }
+  }
+
+  /**
+   * Assigns the specified permissions to the subject with the given origin.
+   * Overwrites all existing permissions, and creates a subject entry if it
+   * doesn't already exist.
+   *
+   * **ATTN:** Assumes that the new permissions have been validated.
+   *
+   * @param origin - The origin of the grantee subject.
+   * @param permissions - The new permissions for the grantee subject.
+   */
+  private setValidatedPermissions(
+    origin: OriginString,
+    permissions: Record<
+      string,
+      ExtractPermission<
+        ControllerPermissionSpecification,
+        ControllerCaveatSpecification
+      >
+    >,
+  ): void {
+    this.update((draftState) => {
+      if (!draftState.subjects[origin]) {
+        draftState.subjects[origin] = { origin, permissions: {} };
+      }
+
+      // Typecast: Immer WritableDraft incompatibility
+      draftState.subjects[origin].permissions = castDraft(permissions);
+    });
+  }
+
+  /**
+   * Constructs the requested caveats for the permission of the specified
+   * subject origin and target name.
+   *
+   * Throws an error if validation fails.
+   *
+   * @see {@link PermissionController.constructCaveat} for details.
+   * @param origin - The origin of the permission subject.
+   * @param target - The permission target name.
+   * @param requestedCaveats - The requested caveats to construct.
+   * @returns The constructed caveats.
+   */
+  private constructCaveats(
+    origin: OriginString,
+    target: ExtractPermission<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >['parentCapability'],
+    requestedCaveats?: unknown[] | null,
+  ): NonEmptyArray<ExtractCaveats<ControllerCaveatSpecification>> | undefined {
+    const caveatArray = requestedCaveats?.map((requestedCaveat) =>
+      this.constructCaveat(origin, target, requestedCaveat),
+    );
+    return caveatArray && isNonEmptyArray(caveatArray)
+      ? caveatArray
+      : undefined;
+  }
+
+  /**
+   * Constructs a caveat for the permission of the specified subject origin and
+   * target name, per the requested caveat argument. This methods validates
+   * everything about the requested caveat, except that its `value` property
+   * is JSON-compatible. It also ensures that a caveat specification exist for
+   * the requested type, and calls the specification validator, if it exists, on
+   * the constructed caveat.
```

Comment:
The note about not validating the JSON no longer applies:
```suggestion
   * everything about the requested caveat. It also ensures that a caveat
   * specification exist for the requested type, and calls the specification
   * validator, if it exists, on the constructed caveat.
```